File: results/0.25/Network-Time-Protocol-Version-4/network-time-protocol-version-4-gemini-1.5-flash.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <arpa/inet.h>

typedef struct {
    uint32_t seconds;
    uint32_t fraction;
} ntp_timestamp_t;

typedef struct {
    uint8_t li_vn_mode;
    uint8_t stratum;
    int8_t poll;
    int8_t precision;
    uint32_t root_delay;
    uint32_t root_dispersion;
    uint32_t reference_id;
    ntp_timestamp_t reference_timestamp;
    ntp_timestamp_t originate_timestamp;
    ntp_timestamp_t receive_timestamp;
    ntp_timestamp_t transmit_timestamp;
    // Add other optional fields as needed...
} ntp_packet_t;

hammer_parser ntp_timestamp_parser() {
    return hammer_map2(
        &(ntp_timestamp_t){},
        hammer_seq(hammer_uint32_be(), hammer_uint32_be())
    );
}

hammer_parser ntp_packet_parser() {
    return hammer_map11(
        &(ntp_packet_t){},
        hammer_seq(hammer_uint8(), hammer_uint8(), hammer_int8(), hammer_int8(), hammer_uint32_be(), hammer_uint32_be(), hammer_uint32_be(), ntp_timestamp_parser(), ntp_timestamp_parser(), ntp_timestamp_parser(), ntp_timestamp_parser())
    );
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <ntp_binary_file>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (fp == NULL) {
        perror("Error opening file");
        return 1;
    }

    fseek(fp, 0, SEEK_END);
    long fsize = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    char *buffer = (char *)malloc(fsize);
    if (buffer == NULL) {
        perror("Memory allocation failed");
        fclose(fp);
        return 1;
    }
    fread(buffer, 1, fsize, fp);
    fclose(fp);

    hammer_result result = hammer_parse(ntp_packet_parser(), buffer, fsize);

    if (result.success) {
        ntp_packet_t packet = *(ntp_packet_t *)result.value;
        printf("NTP Packet:\n");
        printf("LI_VN_Mode: 0x%02X\n", packet.li_vn_mode);
        printf("Stratum: %u\n", packet.stratum);
        printf("Poll: %d\n", packet.poll);
        printf("Precision: %d\n", packet.precision);
        printf("Root Delay: %u\n", packet.root_delay);
        printf("Root Dispersion: %u\n", packet.root_dispersion);
        printf("Reference ID: %u\n", packet.reference_id);
        printf("Reference Timestamp: %u.%u\n", packet.reference_timestamp.seconds, packet.reference_timestamp.fraction);
        printf("Originate Timestamp: %u.%u\n", packet.originate_timestamp.seconds, packet.originate_timestamp.fraction);
        printf("Receive Timestamp: %u.%u\n", packet.receive_timestamp.seconds, packet.receive_timestamp.fraction);
        printf("Transmit Timestamp: %u.%u\n", packet.transmit_timestamp.seconds, packet.transmit_timestamp.fraction);

        free(result.value);
    } else {
        fprintf(stderr, "Error parsing NTP packet: %s\n", result.error);
    }

    free(buffer);
    return 0;
}

File: results/0.25/Network-Time-Protocol-Version-4/network-time-protocol-version-4-gemini-1.5-flash.rs
use nom::{
    bytes::complete::take,
    combinator::{map, map_res},
    number::complete::{be_u32, be_u64, be_u8},
    IResult,
};
use std::fs;
use std::str::from_utf8;

#[derive(Debug)]
struct NtpPacket {
    li: u8,
    vn: u8,
    mode: u8,
    stratum: u8,
    poll: u8,
    precision: u8,
    root_delay: f64,
    root_dispersion: f64,
    reference_id: String,
    reference_timestamp: u64,
    originate_timestamp: u64,
    receive_timestamp: u64,
    transmit_timestamp: u64,
    extension: Vec<u8>,
}

fn parse_ntp_packet(input: &[u8]) -> IResult<&[u8], NtpPacket> {
    let (input, li_vn_mode) = take(1usize)(input)?;
    let li = (li_vn_mode[0] >> 6) as u8;
    let vn = (li_vn_mode[0] >> 3) & 0b111;
    let mode = li_vn_mode[0] & 0b111;

    let (input, stratum) = be_u8(input)?;
    let (input, poll) = be_u8(input)?;
    let (input, precision) = be_u8(input)?;
    let (input, root_delay) = map(be_u32, |x| x as f64 / 65536.0)(input)?;
    let (input, root_dispersion) = map(be_u32, |x| x as f64 / 65536.0)(input)?;
    let (input, reference_id) = map_res(take(4usize), |x| {
        from_utf8(x).map(|id| id.to_string())
    })(input)?;
    let (input, reference_timestamp) = be_u64(input)?;
    let (input, originate_timestamp) = be_u64(input)?;
    let (input, receive_timestamp) = be_u64(input)?;
    let (input, transmit_timestamp) = be_u64(input)?;
    let (input, extension) = take(input.len())(input)?;

    Ok((
        input,
        NtpPacket {
            li,
            vn,
            mode,
            stratum,
            poll,
            precision,
            root_delay,
            root_dispersion,
            reference_id,
            reference_timestamp,
            originate_timestamp,
            receive_timestamp,
            transmit_timestamp,
            extension: extension.to_vec(),
        },
    ))
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <binary_file>", args[0]);
        std::process::exit(1);
    }

    let filename = &args[1];
    let contents = fs::read(filename).expect("Failed to read file");

    match parse_ntp_packet(&contents) {
        Ok((_, packet)) => println!("{:#?}", packet),
        Err(e) => eprintln!("Error parsing NTP packet: {:?}", e),
    }
}
